bibo/internals.py

```diff
@@ -192,3 +195,50 @@ def complete_path(ctx, args, incomplete):
         if os.access(path, os.R_OK):
             options.append((path, os.path.basename(path)))
     return options
+
+
+def match_case(substring, target):
+    """
+    Return the string `substring` with the same case as in the `target`
+    string.
+    """
+    index = target.lower().find(substring.lower())
+    if index < 0:
+        msg = "Failed to match case: {} not in {}".format(substring, target)
+        raise ValueError(msg)
+    return target[index : (index + len(substring))]
+
+
+_ansi_bold = "\033[1m"
+_ansi_unbold = "\033[22m"
+
+
+def highlight_text(text, highlight):
+    """
+    Return `text` with sub-string `highlight` in bold.
+    """
+    text = re.sub(
+        highlight, "{}{}{}".format(_ansi_bold, highlight, _ansi_unbold), text,
+    )
+    # Drop ANSI unbold followed by ANSI bold
+    return text.replace(_ansi_unbold + _ansi_bold, "")
+
+
+def highlight_match(text, result, extra_match_info=None):
+    if extra_match_info is None:
+        extra_match_info = {}
+    for key, vals in result.match.items():
+        if isinstance(vals, collections.abc.Mapping):
+            inner_result = models.SearchResult(result.entry[key], result.match[key],)
+            text, extra_match_info = highlight_match(
+                text, inner_result, extra_match_info,
+            )
+        else:
+            for val in vals:
+                if val in text:
+                    text = highlight_text(text, val)
+                else:
```

Comment:
I'm not sure I get why these are exclusive?

Reply:
If the part of the match already appears in the header / citation (`text` here) you can just highlight that and there's no need to present extra info explaining the match. The `else` block is there to include highlighted info only when it's not part of the `text`. For example, when an entry matches because of something in the abstract.